collect_app/src/main/java/org/odk/collect/android/adapters/IconMenuListAdapter.java

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright 2017 Yura Laguta
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.odk.collect.android.adapters;
+
+import android.content.Context;
+import android.os.Build;
+import android.support.v4.view.ViewCompat;
+import android.view.LayoutInflater;
+import android.view.View;
+import android.view.ViewGroup;
+import android.widget.BaseAdapter;
+import android.widget.TextView;
+
+import org.odk.collect.android.R;
+import org.odk.collect.android.adapters.model.IconMenuItem;
+
+/**
+ * Adapter for List of options with icons
+ */
+public class IconMenuListAdapter extends BaseAdapter {
+
+    private final Context context;
+    private final IconMenuItem[] items;
+
+    public IconMenuListAdapter(Context context, IconMenuItem[] items) {
+        this.context = context;
+        this.items = items;
+    }
+
+    @Override
+    public int getCount() {
+        return items.length;
+    }
+
+    @Override
+    public Object getItem(int position) {
+        return items[position];
+    }
+
+    @Override
+    public long getItemId(int position) {
+        return position;
+    }
+
+    @Override
+    public View getView(int position, View convertView, ViewGroup parent) {
+        if (convertView == null) {
+            convertView = createView(parent);
+        }
+        refreshView((IconMenuItem) getItem(position), (TextView) convertView);
```

Comment:
In addition to checking for null, we should verify that convertView is actually a TextView as well before casting. Ideally this is never really a problem, but the docs <https://developer.android.com/reference/android/widget/Adapter.html#getView(int, android.view.View, android.view.ViewGroup)> recommend it.

Reply:
Very interesting point, it's been a while since I've used listViews but, do you think that is relevant for 'Heterogeneous lists' ? It seems a bit redundant to be and quite defensive, but probably it's better for open source project with many contributors.